src/book/02-system/01-nubots/03-nuclear.mdx

```diff
@@ -5,3 +5,144 @@ title: NUClear
+| IO       | Used to trigger reactions based on standard I/O operations using file descriptors. `on<IO>(pipe/stream/comms, io_status)` will create a task when the communication line matches the given status. For example, `on<IO>(my_file_descriptor, IO::READ)` will create the task when the given file descriptor has data available to read.                                                                                                                                                                                                                                                                |
+| Network  | NUClear provides a networking protocol to send messages to other devices on the network. `on<Network<T>>()` will create a task when T is emitted using the emission type `Scope::NETWORK`.                                                                                                                                                                                                                                                                 |
+
+More information on DSL words for on statements can be found in the [NUClear documentation](https://nuclear.readthedocs.io/en/latest/dsl.html#on-statements).
+
+### Runtime arguments
+
+Some DSL words will provide the ability to make changes to the system during runtime. This means that NUClear avoids the need for a system restart should a configuration, port number, or file need to be changed while the system is running.
+
+The DSL words that take runtime arguments are: `IO`, `TCP`, and `UDP`.
+
+### Callback
+
+The callback which will execute when the reaction is triggered during runtime can be defined using a C++ lambda function.
+
+At runtime, the argument selection for the callback works on the principle of fission, in that the arguments provided to the callback can be deduced and selectively applied as needed. For example:
+
+```cpp
+on<<Trigger<A>, Optional<Trigger<B>>().then([](const A& a, const B& b) {
+
+});
+```
+
+In the above request, the trigger on data type B has been listed as optional, while the trigger for data type A is listed as mandatory. Yet the callback function lists arguments for both A and B.
+
+Let's say that type A is emitted to the PowerPlant, but at this time, type B does not have any data associated with it.
+
+Even though type B is missing data, since it's listed as optional, the task associated with this reaction can be scheduled. When executing the callback for the reaction, NUClear will identify that type B is not present, and will remove reference to this type from the callback, so that the task is only run with type A.
+
+Effectively, through the application of fission, the callback is restructured to look like the following example.
+
+```cpp
+.then([](const A& a){
+
+});
```

Comment:
This isn't really how Fission works and also you'll still get a `b` it's just that if you try to use it you'll get a segmentation fault. I'll go put a more thorough explanation about how optional works and how you should use it up near that DSL word.

What fission actually is, is a way to split up function calls into multiple inputs and is mainly involved in emit from the point of view of the user. So for example, if you call `emit<Scope::NETWORK, Scope::DELAY>(message, "nusight", std::chrono::seconds(1));` the "fission" part, is that NUClear will send the `"nusight"` argument to the network emit and `std::chrono::seconds(1)` to the delay part. (not this won't combine them and network emit in 1 second, it'll just do an emit over the network **AND** a local emit in 1 second.

Closer to what you're describing here is an application of fusion and `RelevantArguments` that is used to ignore output from complicated DSL statements.
For example, when you do `on<Network<Data>, With<Sensors>>`, the network DSL word will make two arguments `const NetworkSource&, const Data&` and the with will make 1 `const Sensors&` but you can call it with `on<Network<Data>>().then([this](const Data& d, const Sensors& s) {});` which will "drop" the NetworkSource argument. This can be useful when you actually don't care about the contents of a message, or don't care about the source of a network packet etc. The matching is a greedy matching with each argument taking as much as it can before moving on.

Reply:
I've changed the callback section if you can have a look. I'm not sure what should be added there, since that section seemed to just talk a lot about the optional word, which is addressed later.